mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

```diff
@@ -2848,6 +2848,41 @@ struct LLVMOpAsmDialectInterface : public OpAsmDialectInterface {
 // DialectInlinerInterface
 //===----------------------------------------------------------------------===//
 
+[[nodiscard]] static bool allocaHasConstantSize(LLVM::AllocaOp allocaOp) {
+  Operation *definingOp = allocaOp.getArraySize().getDefiningOp();
+  return definingOp && definingOp->hasTrait<OpTrait::ConstantLike>();
+}
+
+/// Move all `LLVM::AllocaOp` operations with a constant size in `sourceBlock`
+/// to the end of `entryBlock`. If necessary, the operation defining the
+/// constant will be moved as well.
+static void moveConstantAllocasToEntryBlock(Block *sourceBlock,
+                                            Block *entryBlock) {
+  assert(entryBlock->isEntryBlock() &&
+         "Second argument must be an entry block.");
+  assert(sourceBlock->getParent() == entryBlock->getParent() &&
+         "Blocks must belong to the same region.");
+  if (sourceBlock == entryBlock)
+    // Nothing to do.
+    return;
+  SmallVector<LLVM::AllocaOp> allocasToMove;
+  // Don't walk nested regions of the block, since we can't be sure of their
+  // semantics.
+  for (auto allocaOp : sourceBlock->getOps<LLVM::AllocaOp>()) {
+    if (!allocaHasConstantSize(allocaOp))
+      continue;
+    allocasToMove.push_back(allocaOp);
+  }
+  for (auto allocaOp : allocasToMove) {
+    allocaOp->moveBefore(&entryBlock->back());
+    // If the operation defining the constant array size no longer dominates
+    // the alloca, bring it along to the entry block.
+    Operation *definingOp = allocaOp.getArraySize().getDefiningOp();
+    if (!definingOp->isProperAncestor(allocaOp.getOperation()))
+      definingOp->moveBefore(allocaOp);
```

Comment:
I assume we cannot just create a new constant since we have no builder here? Then we could basically take the matched integer attribute and directly use it here.

Reply:
Is this preferred over moving it (if yes, why?)?

Reply:
There is a lot of underlying assumptions here (e.g. there are no other operations that need to be moved transitively, or the op result is still visible for all users after the move). I think all the assumptions hold here but creating a new constant trivially works and is probably cheaper than figuring out if an op is an ancestor of another operation.

Also isProperAncestor seems suspicious? Isn't this testing if definingOp is a parent of allocaOp? That is always the case. I would assume something like !definingOp dominates allocaOp would be needed?

Reply:
So in brief, it is a lot easier :)

Reply:
Alternatively we can _always_ move the constant, instead of checking anything, as long as we make sure it's before any allocas that we have moved?

Constants cannot have any inputs, so there cannot be anything that needs to be moved transitively, and the visibility in the entry block is always higher than wherever it was before.

Creating a new constant means adding more operations for no particular reason, and they might not even be identical (since in principle the input cut be an `arith.constant`, but we would create an `llvm.mlir.constant`).

What do you think?

Reply:
Always moving to the **beginning** of the block should work as far as I can see. You may also want to check if the constant is already in the entry block of the caller. Then moving again is probably not necessary since we insert the alloca at the end of the entry block.

Reply:
Updated to create a new constant op. Let me know what you think